### ISSUE TITLE
False positives found in Platform Compatibility Analyzer

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/roslyn-analyzers#5962]
### False positives found in platform relations of CA1416

**Diagnostic ID**: [CA1416](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1416): Validate Platform compatibility

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0) or later

These false positives are mainly unhandled edge case scenarios of the [new rules iOS is MacCatalyst](https://github.com/dotnet/runtime/issues/53084#issuecomment-877750424) we added around a year ago. Here we introduced a new concept of **`related platforms`** and below scenarios are not reproduceable for platforms not having such relation.

#### Case 1: False positive when callsite and called API attributes are related and callsite has higher version support than the called API version for `PlatformNeutralAssembly`
```cs
[assembly: SupportedOSPlatform("MacCatalyst13.1")]
public class Test
{
    private static int field = 0;
    [SupportedOSPlatform("ios11.0")]
    public static void iOS11Method() { field = 1; } // Warns for the access to field
}
```
#### Case 2: False positive when API with `[SupportedOSPlatformGuard("ios11.0")]` attribute calls a field within a targeted assembly
```cs
[assembly: SupportedOSPlatform("ios10.0")]
class Test
{
    static readonly bool s_isiOS11OrNewer => OperatingSystem.IsIOSVersionAtLEast(11);

    [SupportedOSPlatformGuard("ios11.0")]
    private bool IsIos11Supported() => s_isiOS11OrNewer; // warns, not expected
    void M1()
    {
        if (IsIos11Supported())
        {
            SupportedOniOS11();
        }
    }
    [SupportedOSPlatform("ios11.0")]
    void SupportedOniOS11() { }
}
```

#### Case 3: False positive when API with `iOS` guard attribute is used within `MacCatalyst` targeted build with lower version
```cs
[assembly: SupportedOSPlatform("MacCatalyst13.1")]
public class Test
{
    [SupportedOSPlatformGuard("ios14.0")]
    internal static bool IsiOS14OrNewer => true;

    [SupportedOSPlatform("ios13.4")]
    public static void iOS13Method() { }

    static void M1()
    {
        iOS13Method(); // Warns, expected
        if (IsiOS14OrNewer)
            iOS13Method(); // Should not warn; This call site is reachable on: 'ios' 14.0 and later, 'MacCatalyst' 13.1 and later. 'Test.iOS13Method()' is only supported on: 'MacCatalyst' 13.4 and later.
            
    }
}
```
Related to and found in https://github.com/dotnet/maui/issues/823

CC @jeffhandley
